### PR TITLE
Networking 2 (Cumulative PR) - Swapping over to new NetworkManager 

### PR DESCRIPTION
**Pull request by BaronGreenback**
The new networkManager working with Jellyfin. Lots of functionality isn't available yet.

This was the minimum amount of changes for a successful compilation and execution. 

Depends on PR #4124 and PR #4253 

To be used in conjunction with Web change jellyfin/jellyfin-web#1979

### CODE REVIEW

**Comment by h1dden-da3m0n** [comment]
Here a working link to the Jellyfin-Web PR [jellyfin-web#1979](https://github.com/jellyfin/jellyfin-web/pull/1979)

/note: if you want to use this link in your PR description:
`[jellyfin-web#1979](https://github.com/jellyfin/jellyfin-web/pull/1979)`

sorry for the nitpicks when it comes to linking, I try to help. (pls tell me if I should stop)

File: Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs [hunk 1]
```diff
@@ -10,6 +10,7 @@
 using MediaBrowser.Model.Configuration;
```
Comment (BaronGreenback) [inline Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs:10]: This change is from PR #4253

File: Jellyfin.Networking/Configuration/NetworkConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,223 @@
+#pragma warning disable CA1819 // Properties should not return arrays
```
Comment (BaronGreenback) [inline Jellyfin.Networking/Configuration/NetworkConfiguration.cs:1]: PR #4124

File: Jellyfin.Api/Controllers/StartupController.cs [hunk 3]
```diff
@@ -3,6 +3,7 @@
 using System.Threading.Tasks;
```
Comment (BaronGreenback) [inline Jellyfin.Api/Controllers/StartupController.cs:3]: PR #4124

File: Jellyfin.Networking/Configuration/NetworkConfigurationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+using Jellyfin.Networking.Configuration;
```
Comment (BaronGreenback) [inline Jellyfin.Networking/Configuration/NetworkConfigurationExtensions.cs:1]: PR #4124

File: Jellyfin.Networking/Configuration/NetworkConfigurationFactory.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+using System.Collections.Generic;
```
Comment (BaronGreenback) [inline Jellyfin.Networking/Configuration/NetworkConfigurationFactory.cs:1]: PR #4124

File: Jellyfin.Networking/Jellyfin.Networking.csproj [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (BaronGreenback) [inline Jellyfin.Networking/Jellyfin.Networking.csproj:1]: PR #4124

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 7]
```diff
@@ -0,0 +1,1280 @@
+#pragma warning disable CA1021 // Avoid out parameters
```
Comment (BaronGreenback) [inline Jellyfin.Networking/Manager/NetworkManager.cs:1]: PR #4124

File: RSSDP/SsdpDevicePublisher.cs [hunk 8]
```diff
@@ -300,17 +301,15 @@ protected override void Dispose(bool disposing)
 
                     foreach (var device in deviceList)
                     {
-                        if (!_sendOnlyMatchedHost ||
-                            _networkManager.IsInSameSubnet(device.ToRootDevice().Address, remoteEndPoint.Address, device.ToRootDevice().SubnetMask))
+                        var root = device.ToRootDevice();
```
Comment (BaronGreenback) [inline RSSDP/SsdpDevicePublisher.cs:304]: Midway code

File: MediaBrowser.Common/Net/NetworkExtensions.cs [hunk 9]
```diff
@@ -45,7 +45,7 @@ public static void AddItem(this NetCollection source, IPObject item)
         /// </summary>
         /// <param name="source">The <see cref="NetCollection"/>.</param>
         /// <returns>Returns a string representation of this object.</returns>
-        public static string Readable(this NetCollection source)
+        public static string AsString(this NetCollection source)
         {
             var sb = new StringBuilder();
```
Comment (h1dden-da3m0n) [inline MediaBrowser.Common/Net/NetworkExtensions.cs:50]: Why create a StringBuilder if you don't use it?

File: MediaBrowser.Common/Net/NetworkExtensions.cs [hunk 10]
```diff
@@ -45,7 +45,7 @@ public static void AddItem(this NetCollection source, IPObject item)
         /// </summary>
         /// <param name="source">The <see cref="NetCollection"/>.</param>
         /// <returns>Returns a string representation of this object.</returns>
-        public static string Readable(this NetCollection source)
+        public static string AsString(this NetCollection source)
```
Comment (h1dden-da3m0n) [inline MediaBrowser.Common/Net/NetworkExtensions.cs:48]: DISCLAIMER: this is the first thing that came to my mind, but i have not checked the code if this is possible.

Why not call this in a overwritten toString or have the whole logic of the method in a overwritten toString?

Comment (BaronGreenback) [reply MediaBrowser.Common/Net/NetworkExtensions.cs:48]: Originally all this code was in the Netcollection class which was an inheritence of Collection. However, creating an inheritence was frowned upon - hence - in an attempt to get it accepted, i moved the code to the network extension class.

File: MediaBrowser.Common/Net/NetworkExtensions.cs (hunk 9 above)
Comment (BaronGreenback) [reply MediaBrowser.Common/Net/NetworkExtensions.cs:50]: Good spot - left behind after an attempt to make a suggested change - and test. The overhead of using a sb with the number of iterations in this function was too big.

File: MediaBrowser.Common/Net/INetworkManager.cs [hunk 11]
```diff
@@ -1,97 +1,233 @@
-#pragma warning disable CS1591
```
Comment (BaronGreenback) [inline MediaBrowser.Common/Net/INetworkManager.cs:1]: This file has been moved.

**Review by hawken93 — COMMENTED** [review]
I have seen through everything, but I don't feel like I would know if something is good or bad. Carry on :+1:

File: Emby.Server.Implementations/LiveTv/TunerHosts/HdHomerun/HdHomerunUdpStream.cs [hunk 12]
```diff
@@ -50,14 +52,35 @@ public class HdHomerunUdpStream : LiveStream, IDirectStreamProvider
...
+            var properties = IPGlobalProperties.GetIPGlobalProperties();
+
+            // Get active udp listeners.
+            var udpListenerPorts = properties.GetActiveUdpListeners()
+                        .Where(n => n.Port >= range.Min && n.Port <= range.Max)
+                        .Select(n => n.Port);
+
+            return Enumerable.Range(range.Min, range.Max)
+                .Where(i => !udpListenerPorts.Contains(i))
+                .FirstOrDefault();
```
Comment (crobibero) [inline Emby.Server.Implementations/LiveTv/TunerHosts/HdHomerun/HdHomerunUdpStream.cs:72]:
```suggestion
            return Enumerable
                .Range(range.Min, range.Max)
                .FirstOrDefault(i => !udpListenerPorts.Contains(i));
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+using Emby.Dlna.PlayTo;
+using Jellyfin.Networking.Configuration;
+using Jellyfin.Networking.Manager;
+using MediaBrowser.Common.Configuration;
+using MediaBrowser.Common.Net;
+using Moq;
+using Microsoft.Extensions.Logging.Abstractions;
+using Xunit;
+using NetCollection = System.Collections.ObjectModel.Collection<MediaBrowser.Common.Net.IPObject>;
+using XMLProperties = System.Collections.Generic.Dictionary<string, string>;
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs:12]:
```suggestion
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs [hunk 14]
```diff
@@ -0,0 +1,485 @@
...
+
+namespace NetworkTesting
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs:14]:
```suggestion
namespace Jellyfin.Networking.Tests
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs [hunk 15]
```diff
@@ -0,0 +1,485 @@
...
+{
+    public class NetTesting
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs:16]:
```suggestion
    public class NetworkParseTests
```

Also fix filename

File: tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs [hunk 16]
```diff
@@ -0,0 +1,485 @@
...
+        [InlineData("localhost:1345")]
+        [InlineData("www.google.co.uk")]
+        [InlineData("fd23:184f:2029:0:3139:7386:67d7:d517")]
+        [InlineData("fd23:184f:2029:0:3139:7386:67d7:d517/56")]
+        [InlineData("[fd23:184f:2029:0:3139:7386:67d7:d517]:124")]
+        [InlineData("fe80::7add:12ff:febb:c67b%16")]
+        [InlineData("[fe80::7add:12ff:febb:c67b%16]:123")]
+        [InlineData("192.168.1.2/255.255.255.0")]
+        [InlineData("192.168.1.2/24")]
+
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs:110]:
```suggestion
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs [hunk 17]
```diff
@@ -0,0 +1,485 @@
...
+        [InlineData("127.0.0.1#")]
+        [InlineData("localhost!")]
+        [InlineData("fd23:184f:2029:0:3139:7386:67d7:d517:1231")]
+        public void TestInvalidCollectionCreation(string address)
+        {
+            Assert.False(TryParse(address, out _));
+        }
+
+        [Theory]
+        // Src, IncIP6, incIP4, exIP6, ecIP4, net
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs:127]: What does this comment mean

File: Jellyfin.Api/Helpers/ClassMigrationHelper.cs [hunk 18]
```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Reflection;
+
+namespace Jellyfin.Api.Migrations
```
Comment (crobibero) [inline Jellyfin.Api/Helpers/ClassMigrationHelper.cs:4]: Why is this in Jellyfin.Api?

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 19]
```diff
@@ -293,6 +302,22 @@ public virtual bool CanLaunchWebBrowser
             ApplicationUserAgent = Name.Replace(' ', '-') + "/" + ApplicationVersionString;
         }
 
+        /// <summary>
+        /// Temporary function to migration network settings out of system.xml and into network.xml.
+        /// TODO: remove at the point when a fixed migration path has been decided upon.
+        /// </summary>
+        private void MigrateNetworkConfiguration()
```
Comment (crobibero) [inline Emby.Server.Implementations/ApplicationHost.cs:309]: Why not just use the migration framework we use for everything else?

File: tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs (hunk 17 above)
Comment (BaronGreenback) [reply tests/Jellyfin.Networking.Tests/NetworkTesting/UnitTesting.cs:125]: I'll expand comment

File: Jellyfin.Api/Helpers/ClassMigrationHelper.cs (hunk 18 above)
Comment (BaronGreenback) [reply Jellyfin.Api/Helpers/ClassMigrationHelper.cs:4]: Left over in error. will correct.

File: Emby.Server.Implementations/ApplicationHost.cs (hunk 19 above)
Comment (BaronGreenback) [reply Emby.Server.Implementations/ApplicationHost.cs:309]: I did try (hence the wrong namespace above) but the migration framework is done too late in the process. 
Networkmanager is created before everything else.

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 20]
```diff
@@ -0,0 +1,482 @@
...
+            var conf = new NetworkConfiguration()
+            {
+                EnableIPV6 = true,
+                EnableIPV4 = true,
+                LocalNetworkSubnets = network.Split(',')
+            };
+
+            using var nm = new NetworkManager(GetMockConfig(conf), new NullLogger<NetworkManager>());
+
+            Assert.True(!nm.IsInLocalNetwork(value));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:93]:
```suggestion
            Assert.False(nm.IsInLocalNetwork(value));
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 21]
```diff
@@ -0,0 +1,482 @@
+using System;
+using System.Net;
+using Emby.Dlna.PlayTo;
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:3]:
```suggestion
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 22]
```diff
@@ -0,0 +1,482 @@
...
+                EnableIPV6 = true,
+                EnableIPV4 = true,
+                LocalNetworkSubnets = lan?.Split(';') ?? throw new ArgumentNullException(nameof(lan))
+            };
+
+            NetworkManager.MockNetworkSettings = interfaces;
+            using var nm = new NetworkManager(GetMockConfig(conf), new NullLogger<NetworkManager>());
+            NetworkManager.MockNetworkSettings = string.Empty;
+
+            Assert.True(string.Equals(nm.GetInternalBindAddresses().AsString(), value, StringComparison.Ordinal));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:72]:
```suggestion
            Assert.Equal(nm.GetInternalBindAddresses().AsString(), value);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 23]
```diff
@@ -0,0 +1,482 @@
...
+
+            using var nm = new NetworkManager(GetMockConfig(conf), new NullLogger<NetworkManager>());
+
+            // Test included, IP6.
+            Collection<IPObject> nc = nm.CreateIPCollection(settings.Split(","), false); 
+            Assert.True(string.Equals(nc?.AsString(), result1, System.StringComparison.OrdinalIgnoreCase));
+
+            // Test excluded, non IP6.
+            nc = nm.CreateIPCollection(settings.Split(","), true);
+            Assert.True(string.Equals(nc?.AsString(), result3, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:176]:
```suggestion
            Assert.Equal(nc.AsString(), result3);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 24]
```diff
@@ -0,0 +1,482 @@
...
+            {
+                EnableIPV6 = true,
+                EnableIPV4 = true,
+            };           
+
+            using var nm = new NetworkManager(GetMockConfig(conf), new NullLogger<NetworkManager>());
+
+            // Test included, IP6.
+            Collection<IPObject> nc = nm.CreateIPCollection(settings.Split(","), false); 
+            Assert.True(string.Equals(nc?.AsString(), result1, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:172]:
```suggestion
            Assert.Equal(nc.AsString(), result1);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 25]
```diff
@@ -0,0 +1,482 @@
...
+            // Test excluded, non IP6.
+            nc = nm.CreateIPCollection(settings.Split(","), true);
+            Assert.True(string.Equals(nc?.AsString(), result3, System.StringComparison.OrdinalIgnoreCase));
+
+            conf.EnableIPV6 = false;
+            nm.UpdateSettings(conf);
+            
+            // Test included, non IP6.
+            nc = nm.CreateIPCollection(settings.Split(","), false);
+            Assert.True(string.Equals(nc?.AsString(), result2, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:183]:
```suggestion
            Assert.Equal(nc.AsString(), result2);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 26]
```diff
@@ -0,0 +1,482 @@
...
+
+            // Test excluded, including IPv6.
+            nc = nm.CreateIPCollection(settings.Split(","), true);
+            Assert.True(string.Equals(nc?.AsString(), result4, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:187]:
```suggestion
            Assert.Equal(nc.AsString(), result4);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 27]
```diff
@@ -0,0 +1,482 @@
...
+
+            conf.EnableIPV6 = true;
+            nm.UpdateSettings(conf);
+
+            // Test network addresses of collection.
+            nc = nm.CreateIPCollection(settings.Split(","), false);
+            nc = nc.AsNetworks();
+            Assert.True(string.Equals(nc?.AsString(), result5, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:195]:
```suggestion
            Assert.Equal(nc.AsString(), result5);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 28]
```diff
@@ -0,0 +1,482 @@
...
+                EnableIPV6 = true,
+                EnableIPV4 = true,
+            };
+
+            using var nm = new NetworkManager(GetMockConfig(conf), new NullLogger<NetworkManager>());
+
+            Collection<IPObject> nc1 = nm.CreateIPCollection(settings.Split(","), false);
+            Collection<IPObject> nc2 = nm.CreateIPCollection(compare.Split(","), false);
+
+            Assert.True(nc1.Union(nc2).AsString() == result);
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:230]:
```suggestion
            Assert.Equal(nc1.Union(nc2).AsString(), result);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 29]
```diff
@@ -0,0 +1,482 @@
...
+            NetworkManager.MockNetworkSettings = string.Empty;
+
+            _ = nm.TryParseInterface(result, out Collection<IPObject>? resultObj);
+
+            if (resultObj != null)
+            {
+                result = ((IPNetAddress)resultObj[0]).ToString(true);
+                var intf = nm.GetBindInterface(source, out int? _);
+
+                Assert.True(string.Equals(intf, result, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:411]:
```suggestion
                Assert.Equal(intf, result);
```

File: tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs [hunk 30]
```diff
@@ -0,0 +1,482 @@
...
+
+            if (nm.TryParseInterface(result, out Collection<IPObject>? resultObj) && resultObj != null)
+            {
+                // Parse out IPAddresses so we can do a string comparison. (Ignore subnet masks).
+                result = ((IPNetAddress)resultObj[0]).ToString(true);
+            }
+
+            var intf = nm.GetBindInterface(source, out int? _);
+
+            Assert.True(string.Equals(intf, result, System.StringComparison.OrdinalIgnoreCase));
```
Comment (crobibero) [inline tests/Jellyfin.Networking.Tests/NetworkTesting/NetworkParseTests.cs:479]:
```suggestion
            Assert.Equal(intf, result);
```

**Review by crobibero — COMMENTED** [review]
I also don't think the test names are properly descriptive, but I'll let someone else suggest changes there

File: Jellyfin.Api/Helpers/ClassMigrationHelper.cs [hunk 31]
```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Reflection;
+
+namespace Jellyfin.Api.Helpers
+{
+    /// <summary>
+    /// A static class for copying matching properties from one object to another.
+    /// TODO: remove at the point when a fixed migration path has been decided upon.
+    /// </summary>
+    public static class ClassMigrationHelper
```
Comment (crobibero) [inline Jellyfin.Api/Helpers/ClassMigrationHelper.cs:10]: I suggest moving this to Jellyfin.Server, potentially under the `Migrations` directory

Comment (BaronGreenback) [reply Jellyfin.Api/Helpers/ClassMigrationHelper.cs:10]: Moved

Comment (BaronGreenback) [reply Jellyfin.Api/Helpers/ClassMigrationHelper.cs:10]: Just tried. Emby.Server.Implementations can't see it and can't add as circular reference.

**Review by crobibero — APPROVED** [review]

**Review by joshuaboniface — APPROVED** [review]

**Comment by joshuaboniface** [comment]
I've tested and found a few minor bugs, but in general it seems to work fine for the core functionality.

Bugs for posterity:

1. Trying to wildcard bind to v6 crashes Kestrel on startup. It doesn't like `[::]`. But setting an explicit v6 addr as the bind address does work.
2. Unchecking the "Enable Auto Discovery:" flag changes the config but does nothing.

That said, I'm happy enough with it to merge and go for rc1.

**Review by joshuaboniface — APPROVED** [review]

**Pull request merged**